### PR TITLE
New `experimental_streamUI()` API

### PR DESCRIPTION
**Pull request by shuding**
This PR reimplements `experimental_streamUI()` on top of the `LanguageModelV1` spec. Ideally, this new API is gonna replace the current `render()` with the following changes:
- `experimental_streamUI()` is async and returns an object with `value` field as the React node, as well as other fields similar to `experimental_streamText()` does.
- The `model` option aligns with [AI core](https://sdk.vercel.ai/docs/ai-core)'s `model` value.
- The `provider` option is removed.
- `functions` is removed in term of `tools`.
- `tools`'s `render()` method is renamed as `generate()`.

`render()` is still kept unchanged, but with a deprecation message.

Some of the code are on top of #1372.

## Remaining issues to make this API stable
- [x] Throw if passing old options to make the migration smoother
- [ ] Better error handling and error boundary UI
- [ ] Figure out a way to stream `text()` node without reconciliation

### CODE REVIEW

File: packages/core/rsc/render/render.tsx [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+      finished = resolvable.promise;
+    }
+
+    const value = renderer(...args);
+    if (
+      value instanceof Promise ||
+      (value &&
+        typeof value === 'object' &&
+        'then' in value &&
+        typeof value.then === 'function')
```
Comment (lgrammel) [inline packages/core/rsc/render/render.tsx:127]: out of curiosity, why is this needed if we already know it's a Promise?

Comment (lgrammel) [reply packages/core/rsc/render/render.tsx:127]: a nevermind, confused the `||` with `&&`

File: packages/core/rsc/render/render.tsx [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        ui.done();
+      } else {
+        handleRender([{ content, done: true }], textRender, ui);
+        await finished;
+        ui.done();
+      }
+    } catch (error) {
+      // During the stream rendering, we don't want to throw the error to the
+      // parent scope but only let the React's error boundary to catch it.
+      ui.error(error);
```
Comment (lgrammel) [inline packages/core/rsc/render/render.tsx:279]: Throwing the error and handling it here breaks the stream. Sometimes, the stream can be mostly good, only small parts fail (like a single tool call). An alternative could be a more granular error handling (catching/handling for each iteration inside the loop)

**Review by lgrammel — APPROVED** [review]

File: packages/core/rsc/render/render.tsx (hunk 2 above)
Comment (shuding) [reply packages/core/rsc/render/render.tsx:279]: Good point! Need to redesign that part a bit. Related: #1349.

**Review by MISTYKITTY — APPROVED** [review]

**Comment by janpawellek** [comment]
I really appreciate this change! It makes everything easier and smoother. I'm looking forward to this being released 👍🏻

**Comment by ap-1** [comment]
The [AI Core pages](https://sdk.vercel.ai/docs/ai-core) do not list `experimental_streamUI()`, while the [Generative UI pages](https://sdk.vercel.ai/docs/api-reference/generative-ui/render) still recommend `render()`. Was `render()` marked as deprecated without this new API having been documented yet?

This PR mentions a few changes, but ignores the change in the `messages` property. What is `ExperimentalMessage[]` and how does it differ from `AIStateItem[]` used in `render()`? Does this new API play with `ai/rsc` functions like `getAIState()`, `useAIState()`, etc., or is it something entirely different?
